File: src/test/java/com/wave/hextractor/util/SMDChecksumUtilsTest.java
package com.wave.hextractor.util;

import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Objects;

import static org.junit.Assert.*;

public class SMDChecksumUtilsTest {

    @Test
    public void checkUpdateMegaDriveChecksum() throws IOException, URISyntaxException {
        Path incorrect = Paths.get(Objects.requireNonNull(getClass().getClassLoader().getResource("files/incorrect_checksum.md")).toURI());
        Path correct = Paths.get(Objects.requireNonNull(getClass().getClassLoader().getResource("files/correct_checksum.md")).toURI());
        File file = File.createTempFile("test", "checkUpdateMegaDriveChecksum");
        file.deleteOnExit();
        Utils.copyFileUsingStream(incorrect.toFile().getAbsolutePath(), file.getAbsolutePath());
        SMDChecksumUtils.checkUpdateMegaDriveChecksum(file.getAbsolutePath());
        byte[] correctBytes = Files.readAllBytes(correct);
        byte[] updatedBytes = Files.readAllBytes(file.toPath());
        assertArrayEquals(correctBytes, updatedBytes);
        Utils.copyFileUsingStream(correct.toFile().getAbsolutePath(), file.getAbsolutePath());
        SMDChecksumUtils.checkUpdateMegaDriveChecksum(file.getAbsolutePath());
        byte[] updatedBytes2 = Files.readAllBytes(file.toPath());
        assertArrayEquals(correctBytes, updatedBytes2);
    }

    @Test
    public void checkUpdateMegaDriveChecksum1() throws IOException, URISyntaxException {
        Path incorrect = Paths.get(Objects.requireNonNull(getClass().getClassLoader().getResource("files/incorrect_checksum.md")).toURI());
        Path correct = Paths.get(Objects.requireNonNull(getClass().getClassLoader().getResource("files/correct_checksum.md")).toURI());
        File file = File.createTempFile("test", "checkUpdateMegaDriveChecksum2");
        file.deleteOnExit();
        Utils.copyFileUsingStream(incorrect.toFile().getAbsolutePath(), file.getAbsolutePath());
        byte[] correctBytes = Files.readAllBytes(correct);
        byte[] updatedBytes = Files.readAllBytes(file.toPath());
        SMDChecksumUtils.checkUpdateMegaDriveChecksum(updatedBytes);
        assertArrayEquals(correctBytes, updatedBytes);
        Utils.copyFileUsingStream(correct.toFile().getAbsolutePath(), file.getAbsolutePath());
        byte[] updatedBytes2 = Files.readAllBytes(file.toPath());
        SMDChecksumUtils.checkUpdateMegaDriveChecksum(updatedBytes2);
        assertArrayEquals(correctBytes, updatedBytes2);
    }

    @Test
    public void getMegaDriveChecksum() throws URISyntaxException, IOException {
        Path path = Paths.get(Objects.requireNonNull(getClass().getClassLoader().getResource("files/correct_checksum.md")).toURI());
        byte[] fileBytes = Files.readAllBytes(path);
        assertEquals(8192, SMDChecksumUtils.getMegaDriveChecksum(fileBytes));
        assertEquals(SMDChecksumUtils.calculateMegaDriveChecksum(fileBytes), SMDChecksumUtils.getMegaDriveChecksum(fileBytes));
    }

    @Test
    public void calculateMegaDriveChecksum() throws IOException, URISyntaxException {
        Path correct = Paths.get(Objects.requireNonNull(getClass().getClassLoader().getResource("files/correct_checksum.md")).toURI());
        byte[] fileBytes = Files.readAllBytes(correct);
       assertEquals(8192, SMDChecksumUtils.calculateMegaDriveChecksum(fileBytes));
    }

    @Test
    public void updateMegadriveChecksum() throws IOException, URISyntaxException {
        Path incorrect = Paths.get(Objects.requireNonNull(getClass().getClassLoader().getResource("files/incorrect_checksum.md")).toURI());
        byte[] fileBytes = Files.readAllBytes(incorrect);
        int originalChecksum = SMDChecksumUtils.getMegaDriveChecksum(fileBytes);
        int calculatedChecksum = SMDChecksumUtils.calculateMegaDriveChecksum(fileBytes);
        SMDChecksumUtils.updateMegadriveChecksum(calculatedChecksum, fileBytes);
        assertNotEquals(originalChecksum, calculatedChecksum);
        assertEquals(calculatedChecksum, SMDChecksumUtils.getMegaDriveChecksum(fileBytes));
    }
}

File: src/test/kotlin/com/wave/hextractor/util/SMDChecksumUtilsTest.kt
package com.wave.hextractor.util

import com.wave.hextractor.util.SMDChecksumUtils.calculateMegaDriveChecksum
import com.wave.hextractor.util.SMDChecksumUtils.checkUpdateMegaDriveChecksum
import com.wave.hextractor.util.SMDChecksumUtils.getMegaDriveChecksum
import com.wave.hextractor.util.SMDChecksumUtils.updateMegaDriveChecksum
import com.wave.hextractor.util.Utils.copyFileUsingStream
import org.junit.Assert
import org.junit.Test
import java.io.File
import java.io.IOException
import java.net.URISyntaxException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*

class SMDChecksumUtilsTest {
    @Test
    @Throws(IOException::class, URISyntaxException::class)
    fun checkUpdateMegaDriveChecksum() {
        val incorrect =
            Paths.get(Objects.requireNonNull(javaClass.classLoader.getResource("files/incorrect_checksum.md")).toURI())
        val correct =
            Paths.get(Objects.requireNonNull(javaClass.classLoader.getResource("files/correct_checksum.md")).toURI())
        val file = File.createTempFile("test", "checkUpdateMegaDriveChecksum")
        file.deleteOnExit()
        copyFileUsingStream(incorrect.toFile().absolutePath, file.absolutePath)
        checkUpdateMegaDriveChecksum(file.absolutePath)
        val correctBytes = Files.readAllBytes(correct)
        val updatedBytes = Files.readAllBytes(file.toPath())
        Assert.assertArrayEquals(correctBytes, updatedBytes)
        copyFileUsingStream(correct.toFile().absolutePath, file.absolutePath)
        checkUpdateMegaDriveChecksum(file.absolutePath)
        val updatedBytes2 = Files.readAllBytes(file.toPath())
        Assert.assertArrayEquals(correctBytes, updatedBytes2)
    }

    @Test
    @Throws(IOException::class, URISyntaxException::class)
    fun checkUpdateMegaDriveChecksum1() {
        val incorrect =
            Paths.get(Objects.requireNonNull(javaClass.classLoader.getResource("files/incorrect_checksum.md")).toURI())
        val correct =
            Paths.get(Objects.requireNonNull(javaClass.classLoader.getResource("files/correct_checksum.md")).toURI())
        val file = File.createTempFile("test", "checkUpdateMegaDriveChecksum2")
        file.deleteOnExit()
        copyFileUsingStream(incorrect.toFile().absolutePath, file.absolutePath)
        val correctBytes = Files.readAllBytes(correct)
        val updatedBytes = Files.readAllBytes(file.toPath())
        checkUpdateMegaDriveChecksum(updatedBytes)
        Assert.assertArrayEquals(correctBytes, updatedBytes)
        copyFileUsingStream(correct.toFile().absolutePath, file.absolutePath)
        val updatedBytes2 = Files.readAllBytes(file.toPath())
        checkUpdateMegaDriveChecksum(updatedBytes2)
        Assert.assertArrayEquals(correctBytes, updatedBytes2)
    }

    @Throws(URISyntaxException::class, IOException::class)
    @Test
    fun megaDriveChecksum() {
        val path = Paths.get(
            Objects.requireNonNull(javaClass.classLoader.getResource("files/correct_checksum.md")).toURI()
        )
        val fileBytes = Files.readAllBytes(path)
        Assert.assertEquals(8192, getMegaDriveChecksum(fileBytes).toLong())
        Assert.assertEquals(
            calculateMegaDriveChecksum(fileBytes).toLong(), getMegaDriveChecksum(fileBytes).toLong()
        )
    }

    @Test
    @Throws(IOException::class, URISyntaxException::class)
    fun calculateMegaDriveChecksum() {
        val correct =
            Paths.get(Objects.requireNonNull(javaClass.classLoader.getResource("files/correct_checksum.md")).toURI())
        val fileBytes = Files.readAllBytes(correct)
        Assert.assertEquals(8192, calculateMegaDriveChecksum(fileBytes).toLong())
    }

    @Test
    @Throws(IOException::class, URISyntaxException::class)
    fun updateMegaDriveChecksum() {
        val incorrect =
            Paths.get(Objects.requireNonNull(javaClass.classLoader.getResource("files/incorrect_checksum.md")).toURI())
        val fileBytes = Files.readAllBytes(incorrect)
        val originalChecksum = getMegaDriveChecksum(fileBytes)
        val calculatedChecksum = calculateMegaDriveChecksum(fileBytes)
        updateMegaDriveChecksum(calculatedChecksum, fileBytes)
        Assert.assertNotEquals(originalChecksum.toLong(), calculatedChecksum.toLong())
        Assert.assertEquals(calculatedChecksum.toLong(), getMegaDriveChecksum(fileBytes).toLong())
    }
}